src/index.js

```diff
@@ -0,0 +1,188 @@
+  const visitor = {
+    JSXElement(path) {
+      path.replaceWith(buildTemplate(path.scope, transformElement(path)));
+    },
+    JSXFragment(path) {
+      path.replaceWith(buildTemplate(path.scope, transformFragment(path)));
+    },
+  };
+
+  /**
+   * @param {import("@babel/traverse").Scope} scope
+   * @param {import("@babel/types").Expression} scope
+   */
+  function buildTemplate(scope, content) {
+    const ref = scope.generateUidIdentifier("templateReference");
+    scope
+      .getProgramParent()
+      .push({ id: t.cloneNode(ref), init: t.objectExpression([]) });
+
+    return template.expression.ast`ESXToken.template(${ref}, ${content})`;
+  }
+
+  function transformElement(path) {
+    /** @type {import("@babel/types").JSXElement} */
+    const node = path.node;
+
+    const jsxElementName = node.openingElement.name;
+    if (t.isJSXNamespacedName(jsxElementName)) {
+      throw path.buildCodeFrameError(
+        "Namespaced JSX elements are not supported."
+      );
+    }
+
+    let factory;
+    let element;
+    if (t.react.isCompatTag(jsxElementName.name)) {
+      factory = "element";
+      element = jsxToString(jsxElementName);
+    } else {
+      factory = "component";
+      element = jsxToJS(jsxElementName);
+    }
+
+    return t.callExpression(
+      t.memberExpression(t.identifier("ESXToken"), t.identifier(factory)),
+      [
+        element,
+        transformAttributesList(path.get("openingElement")),
+        ...path.get("children").map(transformChild).filter(Boolean),
+      ]
+    );
+  }
+
+  function transformFragment(path) {
+    return t.callExpression(
+      t.memberExpression(t.identifier("ESXToken"), t.identifier("factory")),
+      path.get("children").map(transformChild).filter(Boolean)
+    );
+  }
+
+  /**
+   * @param {import("@babel/types").JSXIdentifier | import("@babel/types").JSXMemberExpression} node
+   * @returns {import("@babel/types").Identifier | import("@babel/types").MemberExpression}
+   */
+  function jsxToJS(node) {
+    if (t.isJSXMemberExpression(node)) {
+      return t.inherits(
+        t.memberExpression(jsxToJS(node.object), jsxToJS(node.property)),
+        node
+      );
+    }
+    return t.inherits(t.identifier(node.name), node);
+  }
+
+  /**
+   * @param {import("@babel/types").JSXIdentifier} node
+   * @returns {import("@babel/types").StringLiteral}
+   */
+  function jsxToString(node) {
+    return t.inherits(t.stringLiteral(node.name), node);
+  }
+
+  function transformAttributesList(path) {
+    /** @type {import("@babel/types").JSXOpeningElement} */
+    const node = path.node;
+
+    if (node.attributes.length === 0) return t.nullLiteral();
+
+    let type = "STATIC_TYPE";
+    for (const attr of node.attributes) {
+      if (t.isJSXSpreadAttribute(attr)) {
+        type = "RUNTIME_TYPE";
+        break;
+      } else if (
+        t.isJSXAttribute(attr) &&
+        t.isJSXExpressionContainer(attr.value)
+      ) {
+        type = "MIXED_TYPE";
```

Comment:
break?

Reply:
No, because it could be followed by a spread property.

Reply:
oh ... never mind, if there's spread after it should keep going ... please ignore my previous comment.